checklist/src/main/scala/checklist/Checked.scala

```diff
@@ -0,0 +1,138 @@
+      case Errored(e, w) => err(e, w)
+      case Warned(w, r) => warn(w, r)
+      case Succeeded(r) => succ(r)
+    }
+
+  def recover(func: (NonEmptyList[E], List[W]) => R): Checked[E, W, R] = recoverWith((e, w) => Succeeded(func(e, w)))
+
+  def recoverWith(func: (NonEmptyList[E], List[W]) => Checked[E, W, R]): Checked[E, W, R] =
+    this match {
+      case Errored(errs, warns) => func(errs, warns)
+      case other => other
+    }
+
+  def combine(other: Checked[E, W, R])(implicit semigroup: Semigroup[R]): Checked[E, W, R] = combineWith(other)(semigroup.combine)
+
+  def combineWith[S, T](other: Checked[E, W, S])(f: (R, S) => T): Checked[E, W, T] =
+    this match {
+      case Errored(errs, warns) =>
+        other.fold(
+          (otherErrs, otherWarns) => Errored(errs |+| otherErrs, warns ++ otherWarns),
+          (otherWarns, _) => Errored(errs, warns ++ otherWarns.toList),
+          _ => Errored(errs, warns)
+        )
+      case Warned(warns, result) =>
+        other.fold(
+          (otherErrs, otherWarns) => Errored(otherErrs, warns.toList ++ otherWarns),
+          (otherWarns, otherResult) => Warned(warns |+| otherWarns, f(result, otherResult)),
+          otherResult => Warned(warns, f(result, otherResult))
+        )
+      case Succeeded(result) =>
+        other.fold(
+          Errored.apply _,
+          (warns, otherResult) => Warned(warns, f(result, otherResult)),
+          otherResult => Succeeded(f(result, otherResult))
+        )
+    }
+
+  def map[S](f: R => S): Checked[E, W, S] =
+    fold(
+      Errored.apply _,
+      (w, r) => Warned(w, f(r)),
+      r => Succeeded(f(r))
+    )
+
+  def mapWarnings[X](f: W => X): Checked[E, X, R] =
+    fold(
+      (e, ws) => Errored(e, ws.map(f)),
+      (ws, r) => Warned(ws.map(f), r),
+      Succeeded(_)
+    )
+
+  def mapErrors[F](f: E => F): Checked[F, W, R] =
+    fold(
+      (es, w) => Errored(es.map(f), w),
+      Warned(_, _),
+      Succeeded(_)
+    )
+
+  def flatMap[S](f: R => Checked[E, W, S]): Checked[E, W, S] =
+    fold(
+      Errored(_, _),
+      (_, result) => combineWith(f(result))((_, s) => s),
+      result => combineWith(f(result))((_, s) => s)
+    )
+
+  def value: Option[R] = fold((_, _) => None, (_, r) => Some(r), Some(_))
+  def warnings: List[W] = fold((_, w) => w, (w, _) => w.toList, _ => Nil)
+  def errors: List[E] = fold((errs, _) => errs.toList, (_, _) => Nil, _ => Nil)
+
+  def iorValidated: List[W] Ior Validated[NonEmptyList[E], R] =
+    fold(
+      (es, ws) => Ior.both(ws, Validated.invalid(es)),
+      (ws, result) => Ior.both(ws.toList, Validated.valid(result)),
+      result => Ior.right(Validated.valid(result))
+    )
+
+  def destructured: (List[E], List[W], Option[R]) =
+    fold(
+      (errs, warns) => (errs.toList, warns, None),
+      (warns, result) => (Nil, warns.toList, Some(result)),
+      result => (Nil, Nil, Some(result))
+    )
+
+  def isValid: Boolean = fold((_, _) => false, (_, _) => false, _ => true)
```

Comment:
Copied this from checked syntax. Maybe warning cases should be counted as valid?

Reply:
It probably needs a better method name. I wasn't too happy with `isValid` IIRC but I couldn't think of anything better.

Reply:
What about `hasValue`? As an end-user I probably only really care if there's a value I can get at, or not.